### PR TITLE
Removing some unncessary allocations in LambdaCompiler

### PR DESCRIPTION
**Pull request by bartdesmet**
This addresses issue #11401.

### CODE REVIEW

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.cs [hunk 1]
```diff
@@ -58,7 +58,7 @@ internal sealed partial class LambdaCompiler
         /// </summary>
         private LambdaCompiler(AnalyzedTree tree, LambdaExpression lambda)
         {
-            Type[] parameterTypes = GetParameterTypes(lambda).AddFirst(typeof(Closure));
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.cs]: Note AddFirst would reallocate the array.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.cs [hunk 2]
```diff
@@ -92,20 +92,16 @@ private LambdaCompiler(AnalyzedTree tree, LambdaExpression lambda, MethodBuilder
             var scope = tree.Scopes[lambda];
             var hasClosureArgument = scope.NeedsClosure;
 
-            Type[] paramTypes = GetParameterTypes(lambda);
-            if (hasClosureArgument)
-            {
-                paramTypes = paramTypes.AddFirst(typeof(Closure));
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.cs]: Same remark as above.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Lambda.cs [hunk 3]
```diff
@@ -136,9 +136,33 @@ private void EmitDelegateConstruction(LambdaExpression lambda)
             EmitDelegateConstruction(impl);
         }
 
-        private static Type[] GetParameterTypes(LambdaExpression lambda)
+        private static Type[] GetParameterTypes(LambdaExpression lambda, Type firstType)
         {
-            return lambda.Parameters.Map(p => p.IsByRef ? p.Type.MakeByRefType() : p.Type);
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Lambda.cs]: Change https://github.com/dotnet/corefx/pull/11308 also removes the need for `Map` in `FindMethod` and `FindProperty`. Once both PRs are merged, the `Map` method can likely be removed altogether.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.cs [hunk 4]
```diff
@@ -92,20 +92,16 @@ private LambdaCompiler(AnalyzedTree tree, LambdaExpression lambda, MethodBuilder
...
-            }
+            Type[] paramTypes = GetParameterTypes(lambda, hasClosureArgument ? typeof(Closure) : null);
 
             method.SetReturnType(lambda.ReturnType);
             method.SetParameters(paramTypes);
-            var paramNames = lambda.Parameters.Map(p => p.Name);
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.cs]: No need to allocate an array for this, only to enumerate it below. As mentioned earlier, together with another PR, this reduces the need for `Map`.

**Comment by VSadov** [comment]
LGTM. Fairly simple change.

**Pull request merged**